### ISSUE TITLE
Add backup label to snapshot

### ISSUE THREAD

**Issue by aawsome** [rustic-rs/rustic#292]
See the discussion in https://github.com/restic/restic/pull/3200

**Comment by damoclark** [comment]
[Relevant matching issue for Restic](https://github.com/restic/restic/issues/4026)

**Comment by aawsome** [comment]
@damoclark I have thought a bit about your lable ID and looked at other backup solutions (like borg backup, etc). All together I agree that there are cases where a label might be handy. 

So I propose to additionally add the label to a snapshot and allow to use it to identify snapshots as it is already done with hostname, paths and tags. These fields are so far used:
- to group snapshots, e.g. in the `snapshots` or `prune` command or when looking for the `latest` snapshot
- to allow filtering snapshots (the `--filter*` options)

I've implemented the addition of labels in #295 yesterday. It can be used by specifying `backup --label`
or in the grouping and/or filter options.

So far the paths have been hard-coded to determine the parent in the `backup` command, so I also added the `--group-by` option there in #295. So, using `backup --label xyz --group-by host,label` would search for the parent only based on the current hostname and the label `xyz`.

@damoclark Would you mind to try out #295 if it suits your needs w.r.t. labels? 

There is one thing I still want to think about: So far, the default for grouping still is `host,paths` to get the same behavior as in restic. This can be of course hard-coded in the config file which I would prefer to suggest. But the default is of course an area of discussion...

**Comment by aawsome** [comment]
@damoclark Are you planing to have a look at #295? I'd like to merge that feature soon such that it will be included in the upcoming 0.4.0 release...

**Comment by damoclark** [comment]
Hi @aawsome,

I only just saw your recent message this morning.  I don't have tooling set up for rust development, and so I haven't had a chance to try it yet.  I haven't compiled a rust project before.  

Okay, I just realised that you have a git repo with binaries for the mainline code base.  It looks like this change is merged there, so I'll give that binary a go. :) 

More broadly, I noticed that you have diverged from Restic with the command line interface (switches and options).  Is a goal of Rustic to maintain compatibility with Restic? Does this merged change to Rustic make Rustic repositories incompatible with Restic?  

Longer term, I am curious as to where you see Rustic going and how you plan to position it wrt Restic.  Is it a competitor, or a companion?  What is your goal, or philosophy for its development?  Or are you scratching a personal itch and sharing it with others?

D.

**Comment by aawsome** [comment]
> Is a goal of Rustic to maintain compatibility with Restic?

Yes, the repository. For the CLI I don't have the goal to maintain compatibility. This is anyway not possible as rustic uses different environmental variables and supports config files. There are also enhancements which do not exist in restic.

> Does this merged change to Rustic make Rustic repositories incompatible with Restic?

No, it only adds a field in the JSON for the snapshots. There are already many other fields added by rustic which don't exist in restic.

> Longer term, I am curious as to where you see Rustic going and how you plan to position it wrt Restic. Is it a competitor, or a companion? What is your goal, or philosophy for its development? 

I hope it is both a competitor and a companion. The main philosophy is to add features more quickly than restic is doing. Honestly, I don't think your restic issue will be tackled within the next 2 or three years. I encountered similar things even though I did propose many PRs for additions.
Being a competitor with additional feature may also help restic to more easily adopt those features.

But another goal I had was to implement something serious in rustic to really learn the depth of the language. And to be honest, I'm quite sure that Rust is very superior to Go. One example is how fast I was able to implement rustic and its features. Anyway I think competition in open source solutions is a good thing. On the other side consistency is important for backups, so the goal is to stay compatible with the restic repo format (in the sense that restic should be always able to read/restore from repositories generated by rustic)
